File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/stat/StatVisitor.java
package com.github.jmlparser.stat;

import com.github.javaparser.HasParentNode;
import com.github.javaparser.ast.Modifier;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.NodeList;
import com.github.javaparser.ast.body.MethodDeclaration;
import com.github.javaparser.ast.body.TypeDeclaration;
import com.github.javaparser.ast.expr.*;
import com.github.javaparser.ast.jml.NodeWithJmlTags;
import com.github.javaparser.ast.jml.body.JmlClassExprDeclaration;
import com.github.javaparser.ast.jml.body.JmlFieldDeclaration;
import com.github.javaparser.ast.jml.body.JmlMethodDeclaration;
import com.github.javaparser.ast.jml.body.JmlRepresentsDeclaration;
import com.github.javaparser.ast.jml.clauses.JmlContract;
import com.github.javaparser.ast.jml.doc.*;
import com.github.javaparser.ast.jml.expr.*;
import com.github.javaparser.ast.visitor.GenericVisitorAdapter;
import com.github.javaparser.ast.visitor.VoidVisitorAdapter;
import com.github.javaparser.jml.JmlDocSanitizer;
import lombok.Getter;

import java.util.*;

/**
 * @author Alexander Weigl
 * @version 1 (24.05.22)
 */
public class StatVisitor extends VoidVisitorAdapter<Object> {
    private final List<List<String>> keys;
    @Getter
    private final Map<List<String>, Stat> newlines = new HashMap<>();
    private final ExpressionCosts expressionCosts = new ExpressionCosts();

    public StatVisitor(List<List<String>> keys) {
        this.keys = keys;
    }

    @Override
    public void visit(JmlDocDeclaration n, Object arg) {
        report(n);
    }

    private void report(JmlDocContainer n) {
        for (List<String> keySet : keys) {
            JmlDocSanitizer doc = new JmlDocSanitizer(new TreeSet<>(keySet));
            String san = doc.asString(n.getJmlComments(), false);
            int newlines = newlines(san);
            getStatistics(keySet).addJmlNewlines(newlines);
        }
    }

    @Override
    public void visit(JmlClassExprDeclaration n, Object arg) {
        update(n, this::update);
    }

    private void update(Stat stat, JmlClassExprDeclaration n) {
        Stat.ClassStat classStat = getClassStat(stat, n);
        Stat.ClassStat.CEStat a = classStat.getClassExpressionSpecification(n.getKind().getIdentifier());
        a.numOf++;
        a.sumOfComplexity += n.getInvariant().accept(new ExpressionComplexity(), expressionCosts);
    }

    @Override
    public void visit(JmlRepresentsDeclaration n, Object arg) {
        update(n, this::update);
    }

    private void update(Stat stat, JmlRepresentsDeclaration repr) {
        Stat.ClassStat c = getClassStat(stat, repr);
        c.setNumOfRepresents(c.getNumOfRepresents());
    }

    @Override
    public void visit(JmlDocStmt n, Object arg) {
        report(n);
    }

    @Override
    public void visit(JmlDoc n, Object arg) {
    }

    @Override
    public void visit(JmlDocType n, Object arg) {
        report(n);
    }


    @Override
    public void visit(JmlMethodDeclaration n, Object arg) {
        super.visit(n, arg);
    }


    @Override
    public void visit(MethodDeclaration n, Object arg) {
        super.visit(n, arg);
    }


    @Override
    public void visit(JmlContract n, Object arg) {
        update(n, this::update);
        super.visit(n, arg);
    }


    private Stat getStatistics(List<String> keySet) {
        return newlines.computeIfAbsent(keySet, k -> new Stat());
    }

    private static int newlines(String text) {
        char[] chars = text.toCharArray();
        int n = 0;
        for (char aChar : chars) {
            if (aChar == '\n') {
                n++;
            }
        }
        return n;
    }


    @Override
    public void visit(JmlFieldDeclaration n, Object arg) {
        update(n, this::update);
    }

    interface Update<R> {
        void fn(Stat s, R node);
    }

    public <R extends NodeWithJmlTags<?>> void update(R n, Update<R> update) {
        for (List<String> keySet : keys) {
            if (equal(keySet, n.getJmlTags())) {
                Stat stat = getStatistics(keySet);
                update.fn(stat, n);
            }
        }
    }

    private void update(Stat stat, JmlContract n) {
        Stat.MethodStat mstat = getMethodStat(stat, n);
        mstat.addNumOfContracts(1);
    }

    private static boolean equal(List<String> keySet, NodeList<SimpleName> jmlTags) {
        if (keySet.size() != jmlTags.size()) {
            return false;
        }

        for (int i = 0; i < keySet.size(); i++) {
            if (!keySet.get(i).equals(jmlTags.get(i).getIdentifier())) {
                return false;
            }
        }
        return true;
    }

    private void update(Stat stat, JmlFieldDeclaration n) {
        if (n.getDecl().hasModifier(Modifier.DefaultKeyword.JML_GHOST)) {
            getClassStat(stat, n).addNumOfGhostFields(1);
        } else if (n.getDecl().hasModifier(Modifier.DefaultKeyword.JML_MODEL)) {
            getClassStat(stat, n).addNumOfModelFields(1);
        }
    }

    private Stat.ClassStat getClassStat(Stat stat, HasParentNode<?> n) {
        Optional<TypeDeclaration> tdecl = n.findAncestor(TypeDeclaration.class);
        if (tdecl.isPresent()) {
            final String fqdn = tdecl.get().getFullyQualifiedName().get().toString();
            return stat.getClassStats(fqdn);
        }
        return null;
    }

    private Stat.MethodStat getMethodStat(Stat stat, HasParentNode<Node> n) {
        Optional<TypeDeclaration> tdecl = n.findAncestor(TypeDeclaration.class);
        Optional<MethodDeclaration> mdecl = n.findAncestor(MethodDeclaration.class);
        if (tdecl.isPresent() && mdecl.isPresent()) {
            final String fqdn = tdecl.get().getFullyQualifiedName().get().toString();
            final String sig = mdecl.get().getSignature().asString();
            return stat.getClassStats(fqdn).getMethodStats(sig);
        }
        throw new IllegalArgumentException();
    }


}

class ExpressionComplexity extends GenericVisitorAdapter<Integer, ExpressionCosts> {
    @Override
    public Integer visit(ArrayAccessExpr n, ExpressionCosts arg) {
        return super.visit(n, arg);
    }

    @Override
    public Integer visit(ArrayCreationExpr n, ExpressionCosts arg) {
        return super.visit(n, arg);
    }

    @Override
    public Integer visit(ArrayInitializerExpr n, ExpressionCosts arg) {
        return super.visit(n, arg);
    }

    @Override
    public Integer visit(AssignExpr n, ExpressionCosts arg) {
        return arg.getAssign() + n.getTarget().accept(this, arg) + n.getValue().accept(this, arg);
    }

    @Override
    public Integer visit(BinaryExpr n, ExpressionCosts arg) {
        //TODO distinguish by operator
        return arg.getMinus() + n.getLeft().accept(this, arg) + n.getRight().accept(this, arg);
    }

    @Override
    public Integer visit(UnaryExpr n, ExpressionCosts arg) {
        return super.visit(n, arg);
    }

    @Override
    public Integer visit(LambdaExpr n, ExpressionCosts arg) {
        return super.visit(n, arg);
    }

    @Override
    public Integer visit(CastExpr n, ExpressionCosts arg) {
        return arg.getCast() + n.accept(this, arg);
    }

    @Override
    public Integer visit(CharLiteralExpr n, ExpressionCosts arg) {
        return arg.getCharLiteral();
    }

    @Override
    public Integer visit(ConditionalExpr n, ExpressionCosts arg) {
        return arg.getConditional() + n.getCondition().accept(this, arg)
                + n.getThenExpr().accept(this, arg)
                + n.getElseExpr().accept(this, arg);
    }

    @Override
    public Integer visit(EnclosedExpr n, ExpressionCosts arg) {
        return n.getInner().accept(this, arg);
    }

    @Override
    public Integer visit(IntegerLiteralExpr n, ExpressionCosts arg) {
        return arg.getIntegerLiteral();
    }

    @Override
    public Integer visit(LongLiteralExpr n, ExpressionCosts arg) {
        return arg.getLongLiteral();
    }

    @Override
    public Integer visit(MethodCallExpr n, ExpressionCosts arg) {
        return arg.getMethodCall() + sum(n.getArguments(), arg);
    }

    @Override
    public Integer visit(NameExpr n, ExpressionCosts arg) {
        return arg.getName();
    }

    @Override
    public Integer visit(NullLiteralExpr n, ExpressionCosts arg) {
        return arg.getNullLiteral();
    }

    @Override
    public Integer visit(JmlQuantifiedExpr n, ExpressionCosts arg) {
        return arg.getQuantor() + n.getVariables().size() * arg.getBinderCostsPerVariable() +
                sum(n.getExpressions(), arg);
    }

    private int sum(NodeList<Expression> n, ExpressionCosts arg) {
        return n.stream().mapToInt(it -> it.accept(this, arg)).sum();
    }

    @Override
    public Integer visit(SuperExpr n, ExpressionCosts arg) {
        return 0;
    }

    @Override
    public Integer visit(SwitchExpr n, ExpressionCosts arg) {
        return n.getSelector().accept(this, arg) +
                n.getEntries().stream().mapToInt(it -> sum(it.getLabels(), arg) + 1).sum();
    }

    @Override
    public Integer visit(PatternExpr n, ExpressionCosts arg) {
        return 0;
    }

    @Override
    public Integer visit(InstanceOfExpr n, ExpressionCosts arg) {
        return arg.get_instanceof() + n.getExpression().accept(this, arg);
    }

    @Override
    public Integer visit(JmlLabelExpr n, ExpressionCosts arg) {
        return super.visit(n, arg);
    }

    @Override
    public Integer visit(JmlLetExpr n, ExpressionCosts arg) {
        return arg.getLet() + arg.getBinderCostsPerVariable() * n.getVariables().getVariables().size();
    }

    @Override
    public Integer visit(JmlMultiCompareExpr n, ExpressionCosts arg) {
        return arg.getCompare() * n.getOperators().size();
    }
}

File: tools/stat/src/main/kotlin/io/github/jmltoolkit/stat/StatVisitor.kt
package io.github.jmltoolkit.stat

import com.github.javaparser.ast.CompilationUnit
import com.github.javaparser.ast.Node
import com.github.javaparser.ast.NodeList
import com.github.javaparser.ast.body.*
import com.github.javaparser.ast.expr.*
import com.github.javaparser.ast.jml.NodeWithJmlTags
import com.github.javaparser.ast.jml.body.JmlClassExprDeclaration
import com.github.javaparser.ast.jml.body.JmlFieldDeclaration
import com.github.javaparser.ast.jml.body.JmlMethodDeclaration
import com.github.javaparser.ast.jml.body.JmlRepresentsDeclaration
import com.github.javaparser.ast.jml.clauses.*
import com.github.javaparser.ast.jml.doc.*
import com.github.javaparser.ast.jml.expr.*
import com.github.javaparser.ast.jml.stmt.*
import com.github.javaparser.ast.modules.ModuleDeclaration
import com.github.javaparser.ast.nodeTypes.NodeWithModifiers
import com.github.javaparser.ast.nodeTypes.NodeWithName
import com.github.javaparser.ast.nodeTypes.NodeWithSimpleName
import com.github.javaparser.ast.stmt.LocalClassDeclarationStmt
import com.github.javaparser.ast.visitor.GenericVisitorAdapter
import com.github.javaparser.ast.visitor.VoidVisitorAdapter
import com.github.javaparser.jml.JmlDocSanitizer
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.util.*

/**
 * @author Alexander Weigl
 * @version 1 (24.05.22)
 */
class StatVisitor(
    private val xmlDocument: Document,
    private val keys: List<String>,
    private val expressionCosts: ExpressionCosts
) : VoidVisitorAdapter<Element>() {
    //region Java Stuff
    override fun visit(n: CompilationUnit, arg: Element) {
        super.visit(n, newJavaContext(arg, n.javaClass.getSimpleName(), n.storage.get().fileName))
    }

    override fun visit(n: MethodDeclaration, arg: Element) {
        val arg = newJavaContext(
            arg, n.javaClass.getSimpleName(),
            n.getDeclarationAsString(false, false, true)
        )
        super.visit(n, arg)
    }

    override fun visit(n: ClassOrInterfaceDeclaration, arg: Element) {
        val e = newJavaContext(arg, n)
        super.visit(n, e)
    }

    override fun visit(n: ConstructorDeclaration, arg: Element) {
        var e = newJavaContext(
            arg, n.javaClass.getSimpleName(),
            n.getDeclarationAsString(false, false, true)
        )
        super.visit(n, e)
    }

    override fun visit(n: AnnotationDeclaration, arg: Element) {
        val e = newJavaContext(arg, n)
        super.visit(n, e)
    }

    override fun visit(n: AnnotationMemberDeclaration, arg: Element) {
        super.visit(n, arg)
    }

    override fun visit(n: EnumDeclaration, arg: Element) {
        super.visit(n, newJavaContext(arg, n))
    }

    override fun visit(n: LocalClassDeclarationStmt, arg: Element) {
        super.visit(n, arg)
    }

    override fun visit(n: ModuleDeclaration, arg: Element) {
        super.visit(n, newJavaContext(arg, n))
    }

    private fun newJavaContext(parent: Element, node: NodeWithName<*>): Element =
        newJavaContext(parent, node.javaClass.getSimpleName(), node.nameAsString)

    private fun newJavaContext(parent: Element, node: NodeWithSimpleName<*>): Element =
        newJavaContext(parent, node.javaClass.getSimpleName(), node.nameAsString)

    private fun newJavaContext(parent: Element, simpleName: String, name: String): Element {
        val e = xmlDocument.createElement(simpleName)
        e.setAttribute("name", name)
        parent.appendChild(e)
        return e
    }

    //endregion
    //region plain text reporting
    override fun visit(n: JmlDocDeclaration, arg: Element) {
        reportPlainText(n, arg)
    }

    override fun visit(n: JmlDoc, arg: Element) {
    }

    override fun visit(n: JmlDocStmt, arg: Element) {
        reportPlainText(n, arg)
    }

    override fun visit(n: JmlDocType, arg: Element) {
        reportPlainText(n, arg)
    }

    private fun reportPlainText(n: JmlDocContainer, arg: Element) {
        val doc = JmlDocSanitizer(keys.toSet())
        val san: String = doc.asString(n.jmlComments, false)
        val newlines = newlines(san)
        val e = xmlDocument.createElement("jml-comment")
        arg.appendChild(e)
        e.setAttribute("newlines", "" + newlines)
        e.setAttribute("type", n.javaClass.getSimpleName())
        e.setAttribute("chars", "" + san.length)
    }

    //endregion
    override fun visit(n: JmlClassExprDeclaration, arg: Element) {
        if (active(n)) {
            val e = newElement(arg, n.kind.identifier)
            val expr = getExpressionStat(n.invariant)
            e.appendChild(expr)
            if (n.name.isPresent) {
                e.setAttribute("name", n.name.get().asString())
            }
            setModifierAsAttributes(n, e)
        }
    }

    private fun setModifierAsAttributes(n: NodeWithModifiers<*>, e: Element) {
        for (modifier in n.modifiers) {
            e.setAttribute("has" + modifier.keyword.asString(), "true")
        }
    }

    private fun getExpressionStat(expr: Expression): Element {
        val e = xmlDocument.createElement("expr")
        val costs: Int = expr.accept(ExpressionComplexity(), expressionCosts)
        val numOfVariables = numberOfVariables(expr)
        val length = lengthOf(expr)
        e.setAttribute("complexity", "" + costs)
        e.setAttribute("numOfVariables", "" + numOfVariables)
        e.setAttribute("length", "" + length)

        val map = count(expr)
        map.forEach { k, v -> e.setAttribute(k.simpleName, "" + v) }
        return e
    }

    private fun lengthOf(expr: Expression): Int = expr.toString().length

    private fun numberOfVariables(expr: Expression): Int {
        var sum = 0
        for (childNode in expr.childNodes) {
            if (childNode is NameExpr) sum++
            else if (childNode is Expression && !childNode.getChildNodes().isEmpty()) {
                sum += numberOfVariables(childNode)
            }
        }
        return sum
    }

    private fun newElement(parent: Element, tag: String): Element {
        val e = xmlDocument.createElement(tag)
        parent.appendChild(e)
        return e
    }

    private fun active(n: NodeWithJmlTags<*>): Boolean = equal(keys, n.jmlTags)


    override fun visit(n: JmlRepresentsDeclaration, arg: Element) {
        if (active(n)) {
            val a = newElement(arg, "represents")
            a.setAttribute("name", n.nameAsString)
            setModifierAsAttributes(n, a)
        }
    }

    override fun visit(n: JmlMethodDeclaration, arg: Element) {
        var e = arg
        if (active(n)) {
            e = newJavaContext(e, n.javaClass.getSimpleName(), n.methodDeclaration.nameAsString)
            setModifierAsAttributes(n.methodDeclaration, e)
            super.visit(n.methodDeclaration, e)
        }
    }


    override fun visit(n: JmlContract, arg: Element) {
        if (active(n)) {
            var name = "contract_" + n.range.get().begin.line
            if (n.name.isPresent) {
                name = n.name.get().identifier
            }
            val e = newJavaContext(arg, n.javaClass.getSimpleName(), name)
            e.setAttribute("type", n.type.toString())
            setModifierAsAttributes(n, e)
            e.setAttribute("behavior", "" + n.behavior)
            super.visit(n, e)
        }
    }

    override fun visit(n: JmlExpressionStmt, arg: Element) {
        if (active(n)) {
            val e = newElement(arg, n.kind.jmlSymbol())
            e.appendChild(getExpressionStat(n.expression))
        }
    }

    override fun visit(n: JmlUnreachableStmt, arg: Element) {
        if (active(n)) {
            val e = newElement(arg, "jml-unreachable")
        }
    }

    override fun visit(n: JmlBeginStmt, arg: Element) {
        if (active(n)) {
            newElement(arg, "jml-begin")
        }
    }

    override fun visit(n: JmlEndStmt, arg: Element) {
        if (active(n)) {
            newElement(arg, "jml-end")
        }
    }

    override fun visit(n: JmlGhostStmt, arg: Element) {
        if (active(n)) {
            val e = newElement(arg, "jml-ghost")
            e.setAttribute("statements", "")
            super.visit(n, e)
        }
    }


    override fun visit(n: JmlLabelStmt, arg: Element) {
        if (active(n)) {
            newElement(arg, "jml-label")
        }
    }

    override fun visit(n: JmlSimpleExprClause, arg: Element) {
        val e = newElement(arg, n.kind.jmlSymbol)
        e.appendChild(getExpressionStat(n.expression))
    }

    override fun visit(n: JmlSignalsClause, arg: Element) {
        newElement(arg, n.kind.jmlSymbol)
    }

    override fun visit(n: JmlSignalsOnlyClause, arg: Element) {
        val e = newElement(arg, n.kind.jmlSymbol)
        e.setAttribute("numOfTypes", "" + n.types.size)
    }

    override fun visit(n: JmlOldClause, arg: Element) {
        val e = newElement(arg, n.kind.jmlSymbol)
        e.setAttribute("numOfDecls", "" + n.declarations.variables.size)
    }

    override fun visit(n: JmlMultiExprClause, arg: Element) {
        val e = newElement(arg, n.kind.jmlSymbol)
        for (expression in n.expressions) {
            e.appendChild(getExpressionStat(expression))
        }
    }

    override fun visit(n: JmlCallableClause, arg: Element) {
        val e = newElement(arg, n.kind.jmlSymbol)
    }

    override fun visit(n: JmlForallClause, arg: Element) {
        val e = newElement(arg, n.kind.jmlSymbol)
        e.setAttribute("numVars", "" + n.boundedVariables.size)
    }

    override fun visit(n: JmlRefiningStmt, arg: Element) {
        if (active(n)) {
            val e = newElement(arg, "jml-refining")
        }
    }

    override fun visit(n: JmlClauseIf, arg: Element) {
        super.visit(n, arg)
    }

    override fun visit(n: JmlFieldDeclaration, arg: Element) {
        //update(n, this::update)
    }

    interface Update<R> {
        fun fn(parent: Element, node: R)
    }

    private fun <R : NodeWithJmlTags<*>> update(n: R, update: Update<R>) {
    }

    private fun update(parent: JmlFieldDeclaration, n: JmlFieldDeclaration) {
        /*if (n.getDecl().hasModifier(Modifier.DefaultKeyword.JML_GHOST)) {
            getClassStat(stat, n).addNumOfGhostFields(1);
        } else if (n.getDecl().hasModifier(Modifier.DefaultKeyword.JML_MODEL)) {
            getClassStat(stat, n).addNumOfModelFields(1);
        }
         */
    }


    private fun count(e: Expression): Map<Class<*>, Int> {
        val occCounter: MutableMap<Class<*>, Int> = HashMap()
        val q: ArrayDeque<Node> = ArrayDeque<Node>()
        q.add(e)

        while (!q.isEmpty()) {
            val n: Node = q.pop()
            occCounter.compute(n.javaClass) { k: Class<*>, i: Int? -> i?.let { it + 1 } ?: 1 }
            for (childNode in n.childNodes) {
                if (childNode is Expression) {
                    q.add(childNode)
                }
            }
        }
        return occCounter
    }

    companion object {
        private fun newlines(text: String): Int {
            val chars = text.toCharArray()
            var n = 0
            for (aChar in chars) {
                if (aChar == '\n') {
                    n++
                }
            }
            return n
        }


        private fun equal(keySet: List<String>, jmlTags: NodeList<SimpleName>): Boolean {
            if (keySet.size != jmlTags.size) {
                return false
            }

            for (i in keySet.indices) {
                if (keySet[i] != jmlTags.get(i).identifier) {
                    return false
                }
            }
            return true
        }
    }
}

internal class ExpressionComplexity : GenericVisitorAdapter<Int, ExpressionCosts>() {
    override fun visit(n: ArrayAccessExpr, arg: ExpressionCosts): Int = super.visit(n, arg)

    override fun visit(n: ArrayCreationExpr, arg: ExpressionCosts): Int = super.visit(n, arg)

    override fun visit(n: ArrayInitializerExpr, arg: ExpressionCosts): Int = super.visit(n, arg)

    override fun visit(n: AssignExpr, arg: ExpressionCosts): Int =
        arg.assign + n.target.accept(this, arg) + n.value.accept(this, arg)

    override fun visit(n: BinaryExpr, arg: ExpressionCosts): Int =
        //TODO distinguish by operator
        arg.minus + n.left.accept(this, arg) + n.right.accept(this, arg)

    override fun visit(n: UnaryExpr, arg: ExpressionCosts): Int = super.visit(n, arg)

    override fun visit(n: LambdaExpr, arg: ExpressionCosts): Int = super.visit(n, arg)

    override fun visit(n: CastExpr, arg: ExpressionCosts): Int = arg.cast + n.expression.accept(this, arg)

    override fun visit(n: CharLiteralExpr, arg: ExpressionCosts): Int = arg.charLiteral

    override fun visit(n: ConditionalExpr, arg: ExpressionCosts): Int {
        return arg.conditional + n.condition.accept(this, arg) + n.thenExpr
            .accept(this, arg) + n.elseExpr.accept(
            this, arg
        )
    }

    override fun visit(n: EnclosedExpr, arg: ExpressionCosts): Int = n.inner.accept(this, arg)

    override fun visit(n: IntegerLiteralExpr, arg: ExpressionCosts): Int = arg.integerLiteral

    override fun visit(n: LongLiteralExpr, arg: ExpressionCosts): Int = arg.longLiteral

    override fun visit(n: MethodCallExpr, arg: ExpressionCosts): Int = arg.methodCall + sum(n.arguments, arg)

    override fun visit(n: NameExpr, arg: ExpressionCosts): Int = arg.name

    override fun visit(n: NullLiteralExpr, arg: ExpressionCosts): Int = arg.nullLiteral

    override fun visit(n: JmlQuantifiedExpr, arg: ExpressionCosts): Int =
        arg.quantor + n.variables.size * arg.binderCostsPerVariable + sum(n.expressions, arg)

    private fun sum(n: NodeList<Expression>, arg: ExpressionCosts): Int {
        return n.stream().mapToInt { it -> Objects.requireNonNull(it.accept(this, arg), it.javaClass.getSimpleName()) }
            .sum()
    }

    override fun visit(n: JmlTypeExpr, arg: ExpressionCosts): Int = 1

    override fun visit(n: SuperExpr, arg: ExpressionCosts): Int = 0

    override fun visit(n: SwitchExpr, arg: ExpressionCosts): Int {
        return n.selector.accept(this, arg) + n.entries.stream()
            .mapToInt { it -> sum(it.labels, arg) + 1 }
            .sum()
    }

    override fun visit(n: TypePatternExpr, arg: ExpressionCosts): Int = 0
    override fun visit(n: RecordPatternExpr, arg: ExpressionCosts?): Int = super.visit(n, arg)

    override fun visit(n: BooleanLiteralExpr, arg: ExpressionCosts): Int = 0

    override fun visit(n: InstanceOfExpr, arg: ExpressionCosts): Int = arg.instanceof + n.expression.accept(this, arg)

    override fun visit(n: JmlLabelExpr, arg: ExpressionCosts): Int = super.visit(n, arg)

    override fun visit(n: JmlLetExpr, arg: ExpressionCosts): Int =
        arg.let + arg.binderCostsPerVariable * n.variables.variables.size

    override fun visit(n: JmlMultiCompareExpr, arg: ExpressionCosts): Int = arg.compare * n.operators.size
}